File: cpp/42_trapping_rain_water.cpp
#include <vector>

using namespace std;

class Solution {
 public:
  int trap(vector<int>& height) {
    int left = 0, right = height.size() - 1;
    int left_max = 0, right_max = 0;
    int water = 0;
    while (left <= right) {
      if (left_max < right_max) {
        if (height[left] < left_max) {
          water += left_max - height[left];
        } else {
          left_max = height[left];
        }
        ++left;
      } else {
        if (height[right] < right_max) {
          water += right_max - height[right];
        } else {
          right_max = height[right];
        }
        --right;
      }
    }
    return water;
  }
};

File: rust/42_trapping_rain_water.rs
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let (mut left, mut right) = (0, height.len() - 1);
        let (mut left_max, mut right_max) = (0, 0);
        let mut water = 0;

        while left < right {
            if height[left] < height[right] {
                if height[left] < left_max {
                    water += left_max - height[left];
                } else {
                    left_max = height[left];
                }
                left += 1;
            } else {
                if height[right] < right_max {
                    water += right_max - height[right];
                } else {
                    right_max = height[right];
                }
                right -= 1;
            }
        }

        water
    }
}
